pkg/threescale/backend/backend.go

```diff
@@ -253,11 +257,222 @@ func (b *Backend) localReport(cacheKey string, metrics api.Metrics) {
+// 4. report success and authorization success -> hits += new_auth_hits - (snapshot_hits - to_report_hits ) - actually_reported_hits
+// To summarise, if we have a reporting error, we will not amend our local representation of the remote state.
+// If we have an authorization error we will not amend our local state.
+func (b *Backend) handleFlushCacheUpdate(apps []cachedApp) {
+	for _, cachedApp := range apps {
+		// handle case (1)
+		if cachedApp.reportingErr && cachedApp.authErr {
+			// nothing to do here only continue on
+			// todo add logging
+			continue
+		}
+
+		//now that we have handled the dual error case, we know we need to
+		// read from the cache and modify the entry so we take a write lock
+		app, ok := b.cache.Get(cachedApp.cacheKey)
+		if !ok {
+			continue
+		}
+
+		// handle case (3)
+		if !cachedApp.reportingErr && cachedApp.authErr {
+			app.Lock()
+			// This is necessary when, during flushing, we have reported correctly but failed to authorize
+			// we dont have an updated state from 3scale so in order to avoid re-reporting what we have already
+			// successfully reported. We can achieve this safely by modifying our last know local
+			// representation of the remote state with the deltas we know that 3scale has processed.
+			app.addDeltasToRemoteState(cachedApp.deltas)
+			app.UnlimitedCounter = make(UnlimitedCounter)
+			b.cache.Set(cachedApp.cacheKey, app)
+			app.Unlock()
+			continue
+		}
+
+		// handles cases (2, 4)
+		remoteState := getApplicationFromResponse(cachedApp.authResp).RemoteState
+
+		app.Lock()
+		app.adjustLocalState(cachedApp, remoteState)
+		b.cache.Set(cachedApp.cacheKey, app)
+		app.Unlock()
+	}
+}
+
 // GetPeer returns the hostname of the connected backend
 func (b *Backend) GetPeer() string {
 	return b.client.GetPeer()
 }
 
+// calculateDeltas returns a set of metrics that we can report to 3scale using the lowest period known in the counters
+// for metrics with rate limits attached
+func (a *Application) calculateDeltas() api.Metrics {
+	deltas := make(api.Metrics)
+
+	for metric := range a.LimitCounter {
+		lowestPeriod := a.LimitCounter.getLowestKnownPeriodForMetric(metric)
+		if lowestPeriod != nil {
+			if lastKnownState, ok := a.RemoteState[metric][*lowestPeriod]; ok {
+				toReport := a.LimitCounter[metric][*lowestPeriod].current - lastKnownState.current
+				deltas.Add(metric, toReport)
+			}
+		}
+	}
+
+	// now add the metrics with no limits to the delta
+	for unlimitedMetric, reportWithValue := range a.UnlimitedCounter {
+		deltas[unlimitedMetric] = reportWithValue
+	}
+	return deltas
+}
+
+// addDeltasToRemoteState modifies the remote state with the provided deltas
+func (a *Application) addDeltasToRemoteState(deltas api.Metrics) *Application {
+	for metric, value := range deltas {
+		period := a.RemoteState.getLowestKnownPeriodForMetric(metric)
+
+		if period == nil {
+			continue
+		}
+
+		a.RemoteState[metric][*period].current += value
```

Comment:
In this case we have the data for all the periods we track a metric for, and we have to add the data we reported manually. While we do not need the hierarchy, because the deltas already contain flat usages, we still require the different periods to reflect the additions, that is, if I add 2 hits to, say, minute, and I also have day, I must also add 2 to the day counter.

This should be the reverse of computing the deltas: since the deltas take the difference between what the remote state was and what the current state is by looking at what to report at each period for each metric, we need to do the same (ie. locally reporting, but instead of to the current state, to the remote state).